Passing a parameterized interface as param to a function

**V version:** V 0.2.4 1bbc733
**OS:** Linux lambda 4.19.0-18-amd64 #1 SMP Debian 4.19.208-1 (2021-09-29) x86_64 GNU/Linux

**What did you do?**

I'm trying to create a _generic_ interface for different implementations that share common behavior, and pass the interface as param to a function. For example, take a simple `List` interface such as:

```
interface List<T> {
mut:
  add(e T) bool

}
struct LinkedList<T> {
  // ...
}

fn (mut list LinkedList<T>) add(e T) {
  // ...
}

fn do_list_thing(list List) {    // <--- Error here
  // ...
}

```


**What did you expect to see?**

 
**What did you see instead?**

The compiler issued the following error:
```
lambda :: ~/ws-v/test3 » v -cg test3.v 
/tmp/v_1000/test3.10672805546069593736.tmp.c:1828: error: ',' expected (got "list")
builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```
It works if the interface is not parameterized.

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -8250,6 +8250,25 @@ fn (mut c Checker) fn_decl(mut node ast.FnDecl) {
 						mut v := node.scope.find_var(param.name) or { continue }
 						v.is_auto_heap = true
 					}
+					if info.generic_types.len > 0 && !param.typ.has_flag(.generic)
+						&& info.concrete_types.len == 0 {
+						c.error('generic struct in fn declaration must specify the generic type names, e.g. Foo<T>',
+							param.type_pos)
+					}
+				} else if arg_typ_sym.kind == .interface_ {
+					info := arg_typ_sym.info as ast.Interface
+					if info.generic_types.len > 0 && !param.typ.has_flag(.generic)
+						&& info.concrete_types.len == 0 {
+						c.error('generic interface in fn declaration must specify the generic type names, e.g. Foo<T>',
+							param.type_pos)
+					}
+				} else if arg_typ_sym.kind == .sum_type {
+					info := arg_typ_sym.info as ast.SumType
+					if info.generic_types.len > 0 && !param.typ.has_flag(.generic)
+						&& info.concrete_types.len == 0 {
+						c.error('generic sumtype in fn declaration must specify the generic type names, e.g. Foo<T>',
+							param.type_pos)
+					}
 				}
 			}
 			if c.pref.translated && node.is_variadic && node.params.len == 1 && param.typ.is_ptr() {
@@ -8356,15 +8375,6 @@ fn (mut c Checker) fn_decl(mut node ast.FnDecl) {
 			c.error('missing return at the end of an anonymous function', node.pos)
 		} else if !node.attrs.contains('_naked') {
 			c.error('missing return at end of function `$node.name`', node.pos)
-		}
-	}
-	if node.is_method {
-		sym := c.table.get_type_symbol(node.receiver.typ)
-		if sym.kind == .struct_ {
-			info := sym.info as ast.Struct
-			if info.is_generic && c.table.cur_fn.generic_names.len == 0 {
-				c.error('receiver must specify the generic type names, e.g. Foo<T>', node.method_type_pos)
-			}
 		}
 	}
 	node.source_file = c.file

--- a/vlib/v/checker/tests/generics_fn_arg_type_err.out
+++ b/vlib/v/checker/tests/generics_fn_arg_type_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/generics_fn_arg_type_err.vv:14:23: error: generic interface in fn declaration must specify the generic type names, e.g. Foo<T>
+   12 | }
+   13 |
+   14 | fn do_list_thing(list List) { // <--- Error here
+      |                       ~~~~
+   15 |     // ...
+   16 | }

--- a/vlib/v/checker/tests/generics_fn_arg_type_err.vv
+++ b/vlib/v/checker/tests/generics_fn_arg_type_err.vv
@@ -1,0 +1,19 @@
+interface List<T> {
+mut:
+	add(e T) bool
+}
+
+struct LinkedList<T> {
+	// ...
+}
+
+fn (mut list LinkedList<T>) add(e T) {
+	// ...
+}
+
+fn do_list_thing(list List) { // <--- Error here
+	// ...
+}
+
+fn main() {
+}

--- a/vlib/v/checker/tests/generics_method_receiver_type_err.out
+++ b/vlib/v/checker/tests/generics_method_receiver_type_err.out
@@ -1,4 +1,4 @@
-vlib/v/checker/tests/generics_method_receiver_type_err.vv:6:11: error: receiver must specify the generic type names, e.g. Foo<T>
+vlib/v/checker/tests/generics_method_receiver_type_err.vv:6:11: error: generic struct in fn declaration must specify the generic type names, e.g. Foo<T>
     4 | }
     5 |
     6 | pub fn (x Node) str() string {

